Fix #3972: Added TextViewStyleCheck script

manas-yu opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fix #3972 

This PR adds a script, `TextViewStyleCheck.kt`, which ensures that all `TextView` elements in the layout XML files use centrally managed styles. If any `TextView` does not have a style attribute that starts with `@style/`, an error is reported, indicating the file path where the violation occurs.

`bazel run //scripts:check_textview_styles -- $(pwd)`

To execute the tests, use:
`bazel test //scripts/src/javatests/org/oppia/android/scripts/xml:TextViewStyleCheckTest`

made changes to the `scripts/assets/todo_open_exemptions.textproto`  as the TextviewStyleCheck script was added

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

manas-yu committed:
adding text view style check script

manas-yu committed:
formatting

manas-yu committed:
error print fix

manas-yu commented:
@adhiamboperes PTAL. I've added the `TextViewStyleCheck.kt` script, and it's identifying TextView elements in some layout files that do not have the style attribute (i.e., they are missing the `@style/` reference).

Could you please confirm if this output is appropriate? If there’s an issue with the script or if you’d like any adjustments, please let me know!

Here is the current output:
<details><summary>Files being reported</summary>
<p>

```shell
bazel run //scripts:check_textview_styles -- $(pwd)
INFO: Analyzed target //scripts:check_textview_styles (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //scripts:check_textview_styles up-to-date:
  bazel-bin/scripts/check_textview_styles.jar
  bazel-bin/scripts/check_textview_styles.jdeps
INFO: Elapsed time: 1.388s, Critical Path: 1.09s
INFO: 3 processes: 1 internal, 1 linux-sandbox, 1 worker.
INFO: Build completed successfully, 3 total actions
INFO: Running command line: bazel-bin/scripts/check_textview_styles /home/manas/opensource/oppia-android
TextView Style Check FAILED:
/home/manas/opensource/oppia-android/app/src/main/res/layout-land/question_player_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-land/profile_progress_header.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-land/create_profile_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-land/walkthrough_final_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-port/onboarding_app_language_selection_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-port/question_player_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-port/onboarding_profile_type_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-port/create_profile_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/feedback_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/test_text_view_for_int_with_data_binding.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/beta_notice_dialog_content.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/story_header_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/drawer_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/test_margin_bindable_adapter_activity.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/selection_interaction_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/question_player_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/multiple_choice_interaction_items.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/profile_reset_pin_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/onboarding_language_dropdown_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/lessons_completed_chapter_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/audio_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/app_version_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/state_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/test_text_view_for_live_data_with_data_binding.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/app_language_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/content_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/onboarding_profile_type_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/bottom_sheet_options_menu_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/submitted_answer_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/lessons_in_progress_chapter_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/coming_soon_topic_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/revision_card_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/profile_list_sync_status_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/lessons_locked_chapter_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/license_text_viewer_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/create_profile_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/walkthrough_welcome_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/test_text_view_bindable_adapter_activity.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/story_chapter_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/item_selection_interaction_items.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/profile_picture_edit_dialog.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/audio_language_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/walkthrough_topic_summary_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/walkthrough_final_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/lessons_not_started_chapter_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/promoted_story_card.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/profile_list_learner_id_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/survey_multiple_choice_item.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/general_availability_upgrade_notice_dialog_content.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout/welcome.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/onboarding_app_language_selection_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/question_player_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/onboarding_profile_type_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/create_profile_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/topic_info_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/resume_lesson_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/option_activity.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/options_without_drawer_activity.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/profile_progress_header.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/administrator_controls_activity.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/revision_card_fragment.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/story_chapter_view.xml
/home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/profile_chooser_fragment.xml
Exception in thread "main" java.lang.Exception: Some TextView elements do not have centrally managed styles.
        at org.oppia.android.scripts.xml.TextViewStyleCheckKt.main(TextViewStyleCheck.kt:55)

``` 

</p>
</details>

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu commented:
I need some clarification regarding the issue. The issue states that the script should parse all layout XML (files in `app/src/main/res/layout*`) and ensure all `TextView` elements have a style set (using `@style/`) that aligns with the local codebase style. However, some files (mentioned in the previous comment) have TextView elements without a style attribute. Certain cases need to be considered, such as when the text is center-aligned (`gravity="center"`), uses hardcoded left/right attributes, or lacks direction-sensitive layout attributes. To proceed, I would need an exact set of criteria to differentiate TextView elements with a style set for RTL/LTR settings from those without. This will help ensure the script enforces the style requirement accurately without affecting valid exceptions.

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'develop' into text-view-script

adhiamboperes reviewed (commented):
Thanks @manas-yu!

Overall, this implementation looks good to me. One question though, you had previously raised some edge cases, but I didn’t see them handled in the script. Do you have a plan for that?

adhiamboperes commented:
@Rd4dev, when you have a moment, could you please take a review pass at this PR?

manas-yu committed:
edge cases

manas-yu committed:
formatting

manas-yu committed:
Merge branch 'text-view-script' of https://github.com/manas-yu/oppia-android into text-view-script

manas-yu committed:
class fix

manas-yu committed:
removing redundant checks

manas-yu committed:
checking style

manas-yu committed:
kdoc fix

manas-yu commented:
Hey @adhiamboperes, PTAL.

After looking into the files logged by the previous implementation, I have defined the **minimal required conditions** for the script to pass for the project's current `TextView` usage. Additionally, I have implemented the check for centrally managed styles as outlined in the second point of [this comment](https://github.com/oppia/oppia-android/issues/3972#issuecomment-2557739023). Below are the finalized conditions:

---

## **Allowed TextViews**
- `TextView` elements with a valid `style` attribute referencing a central style (e.g., `@style/YourStyle`).
- Exempt `TextView` elements:
  - Those with `android:gravity="center"`.
  - Those using dynamic visibility (e.g., `android:visibility` containing data binding expressions).
  - Those explicitly defining `android:textSize`.

---

## **Disallowed TextViews**
- `TextView` elements without a `style` attribute or with an invalid/missing style reference.
- `TextView` elements using **legacy directionality attributes**, including:
  - `android:paddingLeft`, `android:paddingRight`
  - `android:layout_marginLeft`, `android:layout_marginRight`
  - `android:layout_toLeftOf`, `android:layout_toRightOf`
- Centrally styled `TextView` elements missing essential **RTL/LTR properties**, such as:
  - `android:textAlignment`
  - `android:gravity`
  - `android:layoutDirection`
  - `android:textDirection`
  - `android:textSize`

---

I will add the test file for the script once the criteria are confirmed. If there are additional suggestions to make the script either more restrictive or more flexible to suit the project’s requirements, please let me know.

Would also love to hear thoughts and suggestions from @Rd4dev. 

Thanks!

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+      if (styleAttribute?.startsWith("@style/") == true) {
+        validateStyle(styleAttribute, idAttribute, filePath)
+      } else {
+        errors.add("$filePath: TextView ($idAttribute) requires central style.")
```
I think including line numbers in error and warning logs could be helpful, especially when views are missing IDs.

Something like: 

```
ERROR: Missing style attribute in file: app/src/main/res/layout/example.xml, line 45. 
WARNING: Hardcoded left/right attribute in file: app/src/main/res/layout/example.xml, line 52. Consider using start/end.
```

Ref: https://github.com/oppia/oppia-android/pull/4128#issuecomment-1077293100

manas-yu replied:
Thank you for pointing this out! I’ve addressed this in the recent changes by using the `Locator` from the SAX parser to track and report line numbers accurately. PTAL at the updated implementation.

Rd4dev replied:
Resolving since following up in another review thread

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+}
+
+private class TextViewStyleCheck(private val repoRoot: File) {
+  private val errors = mutableListOf<String>()
```
Nit - val `errors` could have a more descriptive name, as it's quite generic and could refer to other exceptions too.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+  }
+  // Determines if a TextView is exempt from requiring a centrally managed style.
+  private fun isExemptFromStyleRequirement(element: Element): Boolean {
+    if (element.getAttribute("android:gravity")?.contains("center") == true) return true
```
This allows TextViews with `gravity=center_vertical` and `gravity=center_horizontal` too. 

@adhiamboperes, can you clarify if `gravity=center_vertical` and `gravity=center_horizontal` are also direction-neutral? Also, which of them are useful for RTL/LTR compatibility? It would be helpful to understand the considerations for using `gravity=start` and `gravity=end` as well.

adhiamboperes replied:
- gravity="center_vertical": This is direction-neutral because it only affects the vertical alignment of the content  within its container.
- gravity="center_horizontal": This is not completely direction-neutral because, while "center" remains consistent in both LTR and RTL layouts, combining it with other horizontal alignment values (like start or end) could have different effects depending on the layout direction.

manas-yu replied:
Thank you for the clarification, @Rd4dev and @adhiamboperes. Here’s the finalized implementation and explanation, as I understand it, for handling the `gravity` property.:

#### 1. Direction-Neutrality
- **`gravity=center_vertical`**: Fully direction-neutral as it only affects vertical alignment.
- **`gravity=center_horizontal`**: Neutral when used alone, but not with `start` or `end`, which are direction-sensitive.

#### 2. Considerations for `start` and `end`
- **`gravity=start` and `gravity=end`**: Adapt dynamically to RTL/LTR layouts, unlike `left` and `right`.
#### Implementation
The following function would ensure that mixed cases (e.g., `center|start`) are not marked as direction-neutral:

```kotlin
private fun isGravityDirectionNeutral(element: Element): Boolean {
    val gravity = element.getAttribute("android:gravity") ?: return false
    val gravityValues = gravity.split("|")
    val directionSensitive = setOf("start", "end", "left", "right")
    val directionNeutral = setOf("center", "center_horizontal", "center_vertical")

    return gravityValues.any { it in directionNeutral } &&
           gravityValues.none { it in directionSensitive }
}
```
Let me know if further refinements are needed.

Rd4dev replied:
[Outdated]

Resolving since the requirements were revised to just check the @style attributes.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+ * - path_to_repository_root: The root path of the repository.
+ *
+ * Example:
+ *   bazel run //scripts:check_textview_styles -- $(pwd)
```
Also, this script should be included as part of the CI checks.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+    }
+  }
+  // Determines if a TextView is exempt from requiring a centrally managed style.
+  private fun isExemptFromStyleRequirement(element: Element): Boolean {
+    if (element.getAttribute("android:gravity")?.contains("center") == true) return true
+    if (hasDynamicVisibility(element)) return true
+    if (element.hasAttribute("android:textSize")) return true
```
I’m not entirely clear on the criteria for the exemptions. I saw that `gravity=center` was [mentioned](https://github.com/oppia/oppia-android/issues/3972#issuecomment-2557739023) as an allowed criterion, but could you provide more details on dynamic visibility and text size? Should their inclusion exempt them from checking for centralized styling? Apologies if I missed any requirement description on this.

manas-yu replied:
> I have defined the minimal required conditions for the script to pass for the project's current TextView usage

As mentioned earlier, the current exemption criteria were defined based on the **minimal conditions** required for the script to pass, given the project's current `TextView` usage. Below are the criteria, along with their rationale:

1. ## **Dynamic visibility (`hasDynamicVisibility`)**
   - This exemption applies when the `android:visibility` attribute contains placeholders (e.g., `{}`), indicating that the visibility of the `TextView` is dynamically determined at runtime. In such cases, the visibility can be conditionally changed based on application logic, and applying a centralized style could override its intended behavior.
   - **Impact**: Removing this criterion would cause errors where `TextViews` use dynamic visibility. Below are the affected files and lines :
     - `ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/audio_fragment.xml, line 123`

2. ## **`textSize` attribute**
   - When the `android:textSize` is explicitly set in the XML, it indicates that a specific text size is essential for the layout or design of the `TextView`. Forcing a centralized style could potentially override this critical size setting, which might not be desirable in all cases.
   - **Impact**: Removing this exemption would cause errors when `textSize` is defined directly in the layout XML. Below are the affected files and lines:

<details><summary>Affected files from removing the textSize exemption</summary>
<p>

```
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-land/question_player_fragment.xml, line 22
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-land/question_player_fragment.xml, line 22
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-land/question_player_fragment.xml, line 22
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-land/create_profile_fragment.xml, line 63
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-port/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-port/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-port/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/story_header_view.xml, line 12
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/drawer_fragment.xml, line 73
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/drawer_fragment.xml, line 73
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/multiple_choice_interaction_items.xml, line 36
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/profile_reset_pin_fragment.xml, line 35
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/app_version_fragment.xml, line 17
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/app_version_fragment.xml, line 17
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/state_fragment.xml, line 160
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/app_language_item.xml, line 35
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/lessons_in_progress_chapter_view.xml, line 30
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/coming_soon_topic_view.xml, line 39
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/coming_soon_topic_view.xml, line 39
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/coming_soon_topic_view.xml, line 39
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/lessons_locked_chapter_view.xml, line 31
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/create_profile_fragment.xml, line 62
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/story_chapter_view.xml, line 89
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/item_selection_interaction_items.xml, line 36
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/profile_picture_edit_dialog.xml, line 7
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/profile_picture_edit_dialog.xml, line 7
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/audio_language_item.xml, line 35
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/promoted_story_card.xml, line 55
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/promoted_story_card.xml, line 55
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/profile_list_learner_id_item.xml, line 21
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/profile_list_learner_id_item.xml, line 21
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/profile_list_learner_id_item.xml, line 21
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/survey_multiple_choice_item.xml, line 34
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout/welcome.xml, line 30
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/question_player_fragment.xml, line 24
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/topic_info_fragment.xml, line 28
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/topic_info_fragment.xml, line 28
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/topic_info_fragment.xml, line 28
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp-land/resume_lesson_fragment.xml, line 58
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/option_activity.xml, line 37
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/options_without_drawer_activity.xml, line 37
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/administrator_controls_activity.xml, line 67
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/revision_card_fragment.xml, line 33
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/story_chapter_view.xml, line 130
ERROR: Missing style attribute in file: /home/manas/opensource/oppia-android/app/src/main/res/layout-sw600dp/profile_chooser_fragment.xml, line 46
``` 

</p>
</details> 

I would appreciate @adhiamboperes's input on whether these exemptions are necessary, or if they need to be addressed differently for consistency.

Rd4dev replied:
[Outdated]

Resolving since the requirements were revised to just check the `@style` attributes.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+    if (hasDynamicVisibility(element)) return true
+    if (element.hasAttribute("android:textSize")) return true
+
+    return !hasDirectionalAttributes(element)
```
Is this correct? The elements with the `directionAttributes` should be exempted, right? (as they include the start/end attributes)

Rd4dev replied:
[Outdated], will follow-up with the new implementations as needed.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+    checkForLegacyDirectionality(element, filePath)
+  }
+  // Validate if the referenced style exists and contains necessary RTL/LTR properties.
+  private fun validateStyle(styleAttribute: String, idAttribute: String, filePath: String) {
```
While the check for the existence of the referenced style is helpful, I wonder if it's possible to avoid redundant RTL/LTR settings checks, as they seem to repeat the same element validations multiple times.

Rd4dev replied:
Resolving, as this is revised to be not in the scope of the issue.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    if (errors.isNotEmpty()) {
+      println("\nTextView Style Check FAILED:")
+      errors.forEach { println(it) }
+      throw Exception("Some TextView elements do not have centrally managed styles.")
```
Nit - For consistency in formatting and proper signaling of failures, capitalize the exception message and use `throw Exception`, e.g., `throw Exception("TEXTVIEW STYLE CHECK FAILED.")`

ref: https://github.com/oppia/oppia-android/blob/17f2ef3044e2ee172692d8954b34e91842f35be9/scripts/src/java/org/oppia/android/scripts/xml/StringResourceValidationCheck.kt#L49-L53

Rd4dev reviewed (commented):
@adhiamboperes, sure! And thanks @manas-yu, would it be possible for both to help me better understand the following areas?

I have a few queries regarding the exemptions, aside from gravity=center. Specifically, should dynamic visibility and text sizes be sufficient for ensuring RTL/LTR compatibility in TextViews, as well as with respect to the gravity attributes mentioned in the review comments? I’d appreciate any further clarification on this.

manas-yu committed:
logging line no. and CI integration

manas-yu committed:
formatting

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
test file

manas-yu committed:
Merge branch 'text-view-script' of https://github.com/manas-yu/oppia-android into text-view-script

manas-yu committed:
bazel setup

manas-yu committed:
revert

manas-yu committed:
avoid redundant RTL/LTR settings checks

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'oppia:develop' into text-view-script

manas-yu committed:
revert

manas-yu committed:
adding ids

manas-yu committed:
updating tests

manas-yu committed:
formatting

manas-yu committed:
adding style

manas-yu committed:
kdoc and todo exemp

manas-yu commented:
Hi @adhiamboperes, @Rd4dev,

PTAL, as discussed in the CLaM meeting. I have implemented the script to check for the `@style` attribute only. The list of attribute IDs that need fixing has been added, and I've replaced line numbers with the logging of attribute IDs. This change was made due to the shortcomings of the SAX parser, as mentioned in the article: [SAX Parser](https://eyalsch.wordpress.com/2010/11/30/xml-dom-2/). The article highlights that the parser is not prepared to handle XML entities and ignores comments and processing instructions (e.g., `<?xml version="1.0" encoding="UTF-8" standalone="yes"?>`). Additionally, CDATA sections are undressed and escaped instead of being copied as-is, which does not affect the semantics.

I have fixed the entries that were missing both the `style` and `id` attributes and added a style for the `administrator_controls_text_view` as an example.

Once the script is approved, I will open an issue to track the missing `style` attributes for the text views.

Please let me know if any changes are required or if I missed anything.

Thanks!

manas-yu committed:
line number

manas-yu committed:
formatting

manas-yu committed:
attri list

manas-yu committed:
kdoc

manas-yu commented:
After further research into the SAX Parser, I developed an implementation that logs the closing line number of the TextView element. Please review the script and let me know if it aligns with the requirements.
TextViewStyleCheck script with attribute id logging [here](https://github.com/oppia/oppia-android/blob/8602e7e2d3a4df7fa777a977b7e830e7a6c65dd2/scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt) with line number logging [here](url)

Rd4dev commented on app/src/main/res/layout/drawer_fragment.xml:
```diff
@@ -103,14 +104,10 @@
-            android:layout_gravity="center_vertical"
-            android:layout_marginStart="12dp"
-            android:fontFamily="sans-serif-medium"
+            android:id="@+id/administrator_controls_text_view"
```
Please move the `id` attribute to the top of the view.

Rd4dev replied:
Resolving, my bad! I overlooked the placement.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,289 @@
+      throw Exception("TEXTVIEW STYLE CHECK FAILED")
+    } else if (directionalityWarnings.isEmpty()) {
+      println("TEXTVIEW STYLE CHECK PASSED")
+    }
```
```suggestion
      throw Exception("TEXTVIEW STYLE CHECK FAILED")
    } else {
      println("TEXTVIEW STYLE CHECK PASSED")
    }
```

We may not need to check for an empty `directionalityWarnings`, as it would prevent the 'PASSED' message from being printed if any warnings are present.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,289 @@
+        doc = docBuilder.newDocument()
+      } catch (e: Exception) {
+        throw RuntimeException("Can't create SAX parser / DOM builder.", e)
+      }
```
We can use `error` - https://kotlinlang.org/docs/exceptions.html#error-function instead of a generic `RuntimeException`, which throws an `IllegalStateException` - close to represent the situation.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,289 @@
+        throw RuntimeException("Can't create SAX parser / DOM builder.", e)
+      }
+
+      val elementStack = Stack<Element>()
```
We can replace the `Stack` functionalities with [`ArrayDeque`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/-array-deque/), to avoid legacy baggage and obtain better performance.

```
push -> addLast
pop -> removeLast
peek -> last
```

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,289 @@
+          attributes: Attributes
+        ) {
+          addTextIfNeeded()
+          val el = doc.createElement(qName)
```
```suggestion
          val element = document.createElement(qName)
```

Suggestions for a more descriptive names.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,289 @@
+
+        override fun endElement(uri: String, localName: String, qName: String) {
+          addTextIfNeeded()
+          val closedEl = elementStack.pop()
```
```suggestion
          val closedElement = elementStack.pop()
```

same as above, suggestion for a more descriptive name.

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,289 @@
+          for (i in 0 until attributes.length) {
+            el.setAttribute(attributes.getQName(i), attributes.getValue(i))
+          }
+          el.setAttribute(lineNumAttribName, locator.lineNumber.toString())
```
Didn't realize earlier that the [lineNumber returns the event ends](https://docs.oracle.com/javase/6/docs/api/org/xml/sax/Locator.html#:~:text=getPublicId()-,getLineNumber,delimited%20by%20line%20ends%2C%20which%20are%20defined%20in%20the%20XML%20specification,-.) - this is tricky and might be confusing to point the line numbers to closing tags.

So, I tried out to see if we can work around this to acquire the opening tag line number by subtracting the length of the attributes from the end events line number and could get the correct opening tag line numbers as far as I have tested. Would like to know your thoughts on this.

```
val lineNumber = element.getAttribute(LINE_NUMBER_ATTRIBUTE).toString()

element.setAttribute(lineNumAttribName, (locator.lineNumber - attributes.length).toString())
```

manas-yu replied:
Thank you for taking this up! I had a similar idea in mind at one point, but it was based on a few assumptions. After reviewing the current format of `TextView`s, I think this approach can work well. Here are the key assumptions:

- Each attribute is assumed to occupy exactly one line.
- Comments or whitespace may exist between attributes.
- Some XML files might have attributes on the same line as the opening tag.

Given these considerations, this approach seems effective for now, and I’ve incorporated it into the implementation.

Rd4dev replied:
Agree, as we don’t have any strict styling checks for the mentioned conditions, any of the failure assumptions might lead to inaccurate line numbers. I think, until a proper fix for handling the presence of the ID (or other related issue) is in place, we can conditionally add the ID information. 

Something like:
```kotlin
val idAttribute = element.attributes.getNamedItem("android:id")?.nodeValue

if (styleAttribute.isNullOrBlank()) {
    val idInformation= idAttribute?.let { " ID: $it " } ?: ""
    // similarly for warnings
    styleValidationIssues.add(
        "ERROR: Missing style attribute in file: $filePath ${idInformation},line $lineNumber."
    )
}
```

The line numbers could be helpful when the ID information is not available, while the ID would be useful when present. I'd also like to hear from @adhiamboperes on the need for tracking the unavailability of the ID attribute.

adhiamboperes replied:
Per the Office hours meeting discussion last week, I seem to recall that we decided to prioritize ids, and the line number of the id such that if the id is missing for an existing textview, we would flag that and fix it(in this PR if the work is small, or other PR if the list is long).  Please see here: https://docs.google.com/document/d/16_GUgViYaggDI5EKUzxFKqj5NbP_SxyvzmaaxJAK_mM/edit?tab=t.0

manas-yu replied:
Thank you, @adhiamboperes, for the clarification! During our earlier discussion, I thought the priority was on adding IDs because the line number logging had some issues and inconsistencies. However, since the line number logging is now working fine, I just wanted to ensure we were aligned.

That said, I’ve added IDs for all the TextViews that were missing them. As discussed, there’s no downside to adding IDs, and it ensures better consistency across the codebase.

adhiamboperes replied:
@Rd4dev, @manas-yu, it is not required for this issue to add a check for ensuring IDs are present, however it may be a nice to have. Perharps creating a new issue to track that would be better, including as much etail as possible. Thanks!

Rd4dev commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,289 @@
+}
+
+// Known exceptions that currently lack a style and are being tracked for fixes.
+private val attributeIds = listOf(
```
Could you share any requirements (if any) discussed in the CLaM meeting regarding the existing issues? I noticed you mentioned the need to add and track an issue.

Was there a suggestion to exempt these issues, or has a fix been proposed? If exempting was considered, would it be helpful to mark them as warnings to ensure they aren't overlooked?

manas-yu replied:
After a discussion with (@)BenHenning, it was decided that the script should identify `TextView` elements missing a `style` attribute. For `TextView` elements currently lacking a `style`, their `id's` should be tracked in an issue for future resolution. Additionally, these `id's` will be hardcoded in the script as a temporary measure. For example, I have assigned the style `AdministratorControlsText` to `@+id/administrator_controls_text_view`.

Rd4dev replied:
Got it! then will just keep this comment unresolved as a reminder to add the Todo for existing `TextView`s lacking a style and will resolve it once done.

Rd4dev reviewed (commented):
Thanks @manas-yu, for the reimplementations with the revised requirements – this is nice! Just a few changes to be made, and I'd like to hear your thoughts on the workaround for line numbers and exemption cases. 

(Note: I haven't had a chance to look at the tests yet, but I'll review them as soon as possible.)

manas-yu committed:
Deque and line no

manas-yu committed:
formatting

manas-yu commented:
Hi @Rd4dev PTAL. Thanks!

Rd4dev commented on scripts/src/javatests/org/oppia/android/scripts/xml/TextViewStyleCheckTest.kt:
```diff
@@ -0,0 +1,326 @@
+
+    createLayoutFile(invalidLayout)
+
+    val thrown = catchThrowable { runScript() }
```
1. Here and elsewhere please replace `catchThrowable` implementation with `assertThrows` as it provides a concise syntax that will fail if no exception is thrown while providing detailed error messages making it easier to debug.

2. Use specific types of exception whenever possible

ref: https://github.com/oppia/oppia-android/blob/develop/scripts/src/javatests/org/oppia/android/scripts/wiki/WikiTableOfContentsCheckTest.kt#L103-L107

Rd4dev replied:
@manas-yu, also checking if the point 2 was addressed - could you ensure the exact type of exception than a generic one. As it is generally good to test for specific cases : wiki reference - https://github.com/oppia/oppia-android/wiki/Writing-Tests-With-Good-Behavioral-Coverage#1-verifying-correct-exception-types 

sample: https://github.com/oppia/oppia-android/blob/8ce374bc30570ceea1973d14eb5c20f65e619649/scripts/src/javatests/org/oppia/android/scripts/wiki/WikiTableOfContentsCheckTest.kt#L104

Rd4dev commented on scripts/src/javatests/org/oppia/android/scripts/xml/TextViewStyleCheckTest.kt:
```diff
@@ -0,0 +1,326 @@
+    assertThat(thrown).hasMessageThat().contains(TEXTVIEW_STYLE_CHECK_FAILED_OUTPUT_INDICATOR)
+    val output = outContent.toString()
+    assertThat(output).contains("WARNING: Hardcoded left/right attribute")
+    assertThat(output).contains("ERROR: Missing style attribute")
+  }
+  @Test
```
Nit: Here and elsewhere, add a new line after each test case.

Rd4dev reviewed (commented):
Thanks @manas-yu, just few test changes and need to conditionally add the id information and the mentioned Todo. Once that's done this should be good to go!

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
TODO and idInformation

manas-yu committed:
formatting

manas-yu committed:
foramtting

manas-yu commented:
Hi @Rd4dev, thank you for the detailed reviews! I’ve added the TODO comment and made the requested changes PTAL.

manas-yu committed:
idinfo logging

adhiamboperes commented on scripts/src/javatests/org/oppia/android/scripts/xml/TextViewStyleCheckTest.kt:
```diff
@@ -0,0 +1,331 @@
+
+    createLayoutFile(invalidLayout)
+
+    val thrown = assertThrows<Exception>() { runScript() }
```
Nit, here and below, but I feel like good variable names are important.
```suggestion
    val exception = assertThrows<Exception>() { runScript() }
```

adhiamboperes reviewed (approved):
Thanks @manas-yu for cleaning up this implementation, and thanks @Rd4dev for the thorough review!

This PR LGTM except for the comment I have. PTAL @manas-yu and reassign.

manas-yu committed:
renaming

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
formatting

manas-yu committed:
formatting

manas-yu committed:
Merge branch 'text-view-script' of github.com:manas-yu/oppia-android into text-view-script

Rd4dev commented on scripts/src/javatests/org/oppia/android/scripts/xml/TextViewStyleCheckTest.kt:
```diff
@@ -0,0 +1,331 @@
+    main(tempFolder.root.absolutePath)
+  }
+
+  private fun catchThrowable(executable: () -> Unit): Throwable {
```
@manas-yu just checking is this `catchThrowable` implementation still used elsewhere?

Rd4dev reviewed (commented):
Thanks, @manas-yu, for the work! The PR is nearly complete..., with just three open comments to address:

1. To specify the type of exception explicitly.
2. Review the `catchThrow` implementation.
3. @adhiamboperes, could you check this comment - https://github.com/oppia/oppia-android/pull/5599/files/c80cfe923bf2fc434ee5151bcbf94846d9101811#r1923867550 and suggest if there's a need to verify the presence of the ID (or track them as issue)?

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
exception handling

manas-yu committed:
Merge branch 'text-view-script' of github.com:manas-yu/oppia-android into text-view-script

manas-yu commented:
Thank you @adhiamboperes @Rd4dev for all the guidance and apologies for all the silly mistakes, I'm still learning the code style and writing good clean code and its great to learn something new with every PR! :)

Rd4dev commented:
> Thank you @adhiamboperes @Rd4dev for all the guidance and apologies for all the silly mistakes

Not at all, @manas-yu, you’re doing fantastic! I’m still learning too. 

Regarding the PR, I think everything is resolved — thanks for the work again! I guess the only remaining comment is about getting advice on handling the missing IDs. @adhiamboperes, please feel free to resolve that if it’s blocking the merge once addressed.

manas-yu committed:
adding id attributes

manas-yu commented:
@adhiamboperes PTAL, Thanks!

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
Merge branch 'develop' into text-view-script

adhiamboperes commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,292 @@
+    if (styleAttribute.isNullOrBlank()) {
+      val idInformation = idAttribute?.let { " ID: $it," } ?: ""
+      styleValidationIssues.add(
+        "ERROR: Missing style attribute in file: $filePath,$idInformation line $lineNumber."
```
```suggestion
        "ERROR: Missing style attribute in file: $filePath, $idInformation line $lineNumber."
```

adhiamboperes commented on scripts/src/java/org/oppia/android/scripts/xml/TextViewStyleCheck.kt:
```diff
@@ -0,0 +1,292 @@
+    }
+  }
+
+  private fun readXMLWithLineNumbers(inputStream: FileInputStream, lineNumAttribName: String):
```
```suggestion
  private fun readXmlWithLineNumbers(inputStream: FileInputStream, lineNumAttribName: String):
```

adhiamboperes reviewed (commented):
Thanks @manas-yu! This PR should be ready to merge after the cleanups I have highlighted.

manas-yu committed:
Merge branch 'develop' into text-view-script

manas-yu committed:
edits

manas-yu commented:
Hi @adhiamboperes PTAL, Thanks!

adhiamboperes reviewed (approved):
LGTM!
